Source/AwsCommonRuntimeKit/crt/CBOR.swift

```diff
@@ -0,0 +1,360 @@
+                guard
+                    aws_cbor_decoder_pop_next_float_val(self.rawValue, &out_value)
+                        == AWS_OP_SUCCESS
+                else {
+                    throw CommonRunTimeError.crtError(.makeFromLastError())
+                }
+                return .double(out_value)
+            }
+        case AWS_CBOR_TYPE_BYTES:
+            do {
+                var out_value: aws_byte_cursor = aws_byte_cursor()
+                guard
+                    aws_cbor_decoder_pop_next_bytes_val(self.rawValue, &out_value)
+                        == AWS_OP_SUCCESS
+                else {
+                    throw CommonRunTimeError.crtError(.makeFromLastError())
+                }
+                return .bytes(out_value.toData())
+            }
+        case AWS_CBOR_TYPE_TEXT:
+            do {
+                var out_value: aws_byte_cursor = aws_byte_cursor()
+                guard
+                    aws_cbor_decoder_pop_next_text_val(self.rawValue, &out_value)
+                        == AWS_OP_SUCCESS
+                else {
+                    throw CommonRunTimeError.crtError(.makeFromLastError())
+                }
+                return .text(out_value.toString())
+            }
+        case AWS_CBOR_TYPE_BOOL:
+            do {
+                var out_value: Bool = false
+                guard
+                    aws_cbor_decoder_pop_next_boolean_val(self.rawValue, &out_value)
+                        == AWS_OP_SUCCESS
+                else {
+                    throw CommonRunTimeError.crtError(.makeFromLastError())
+                }
+                return .bool(out_value)
+            }
+        case AWS_CBOR_TYPE_NULL:
+            do {
+                guard
+                    aws_cbor_decoder_consume_next_single_element(self.rawValue)
+                        == AWS_OP_SUCCESS
+                else {
+                    throw CommonRunTimeError.crtError(.makeFromLastError())
+                }
+                return .null
+            }
+        case AWS_CBOR_TYPE_TAG:
+            var out_value: UInt64 = 0
+            guard
+                aws_cbor_decoder_pop_next_tag_val(self.rawValue, &out_value)
+                    == AWS_OP_SUCCESS
+            else {
+                throw CommonRunTimeError.crtError(.makeFromLastError())
+            }
+            guard
+                out_value == 1
+            else {
+                throw CommonRunTimeError.crtError(CRTError(code: AWS_ERROR_CBOR_UNEXPECTED_TYPE.rawValue))
+            }
+            let timestamp = try popNext()
+
+            if case .double(let value) = timestamp {
+                return .date(Date.init(timeIntervalSince1970: value))
+            } else if case .uint64(let value) = timestamp {
+                return .date(Date.init(timeIntervalSince1970: Double(value)))
+            } else if case .int(let value) = timestamp {
+                return .date(Date.init(timeIntervalSince1970: Double(value)))
```

Comment:
in C, converting from unsigned 64 or int 64 to double may loss the precision. Not sure about swift double, maybe worth to check. But seems like the `timeIntervalSince1970` in swift only takes double, so, maybe just document it out that's a risk of loss of precision?

Reply:
Thanks, I have added a comment.